File: DatabindingWithNewsApi/app/src/main/java/com/enpassio/databindingwithnewsapi/data/NewsRepository.java
package com.enpassio.databindingwithnewsapi.data;

import android.os.AsyncTask;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.enpassio.databindingwithnewsapi.model.Article;
import com.enpassio.databindingwithnewsapi.utils.NetworkUtils;

import java.util.List;

public class NewsRepository {

    private static volatile NewsRepository sInstance;
    private final MutableLiveData<List<Article>> articles = new MutableLiveData<>();
    private static final String TAG = "NewsRepository";

    private NewsRepository() {
        Log.d(TAG, "New instance created");
    }

    public static NewsRepository getInstance() {
        if (sInstance == null) {
            synchronized (NewsRepository.class) {
                if (sInstance == null) {
                    sInstance = new NewsRepository();
                }
            }
        }
        return sInstance;
    }

    public void startFetching() {
        //Start fetching from the News Api in a background thread
        new NewsAsyncTask().execute();
    }

    public LiveData<List<Article>> getArticles() {
        return articles;
    }

    private class NewsAsyncTask extends AsyncTask<Void, Void, List<Article>> {

        @Override
        protected List<Article> doInBackground(Void... voids) {
            return NetworkUtils.fetchArticles();
        }

        @Override
        protected void onPostExecute(List<Article> list) {
            if(list != null && !list.isEmpty()){
                articles.setValue(list);
            }
        }

    }
}

File: DatabindingWithNewsApiKotlin/app/src/main/java/com/enpassio/databindingwithnewsapikotlin/data/NewsRepository.kt
package com.enpassio.databindingwithnewsapikotlin.data

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.os.AsyncTask
import android.util.Log
import com.enpassio.databindingwithnewsapikotlin.model.Article
import com.enpassio.databindingwithnewsapikotlin.utils.fetchArticles


object NewsRepository{

    /*We need a mutable live data here, so that we can modify it, but we
    prefer to pass an immutable live data to the UI layer*/
    private val _articles = MutableLiveData<List<Article>?>()
    val articles: LiveData<List<Article>?>
        get() = _articles

    fun startFetching() {
        //Start fetching from the News Api in a background thread
        NewsAsyncTask().execute()
    }

    private class NewsAsyncTask : AsyncTask<Void, Void, List<Article>>() {

        override fun doInBackground(vararg voids: Void): List<Article>? {
            return fetchArticles()
        }

        override fun onPostExecute(list: List<Article>?) {
            if(list?.isNotEmpty() == true){
                _articles.value = list
                Log.d(TAG, "list size: " + list.size)
            }
        }
    }

    private const val TAG = "NewsRepository"
}
